File: app/src/main/java/br/com/menuvem/lojista/domain/model/ProdutoComponenteCompleto.kt
package br.com.menuvem.lojista.domain.model

/**
 * Componente aplicado a um produto, enriquecido com os dados do componente e
 * seus itens (com insumo) para exibição e cálculo de custo no produto.
 *
 * @param custo Custo da parcela deste componente no produto =
 *        soma dos itens (por inteiro) × multiplicador do vínculo.
 */
data class ProdutoComponenteCompleto(
    val vinculo: ProdutoComponente,
    val componente: Componente,
    val itens: List<ItemComponenteComInsumo>
) {
    val custo: Double
        get() = itens.sumOf { it.custo } * vinculo.multiplicador

    val quantidadeItens: Int
        get() = itens.size
}

File: lib/domain/model/produto_componente_completo.dart
import 'componente.dart';
import 'item_componente_com_insumo.dart';
import 'produto_componente.dart';

/// Componente aplicado a um produto, enriquecido com os dados do componente
/// e seus itens (com insumo) para exibição e cálculo de custo no produto.
///
/// [custo]: custo da parcela deste componente no produto = soma dos itens
/// (por inteiro) × multiplicador do vínculo.
class ProdutoComponenteCompleto {
  final ProdutoComponente vinculo;
  final Componente componente;
  final List<ItemComponenteComInsumo> itens;

  const ProdutoComponenteCompleto({
    required this.vinculo,
    required this.componente,
    required this.itens,
  });

  double get custo =>
      itens.fold(0.0, (sum, item) => sum + item.custo) * vinculo.multiplicador;

  int get quantidadeItens => itens.length;
}
